### ISSUE TITLE
benchmarking qupath-wsinfer

### ISSUE THREAD

**Issue by kaczmarj** [qupath/qupath-docs#89]
i am using this issue to provide info on the data i use when benchmarking wsinfer in qupath.

image: TCGA-3C-AALI-01Z-00-DX1.F6E9A5DF-D8FB-45CF-B4BD-C6B76294C291.svs ([link to image for download](https://portal.gdc.cancer.gov/files/d46167af-6c29-49c7-95cf-3a801181aca4) on Genomic Data Commons)

geojson representation of the square region. the region has an area of 100 millimeters squared. (the geojson coordinates are in pixels, and it is the equivalent of 100 mm2 in the whole slide image).

i will update this issue with my running times for a 12th gen i5 cpu and an nvidia 2080ti gpu, both in windows 11.

```json
{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "id": "9724d4e3-cd6d-4630-a474-4d71e35db065",
      "geometry": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              25471,
              25869
            ],
            [
              65471,
              25869
            ],
            [
              65471,
              65869
            ],
            [
              25471,
              65869
            ],
            [
              25471,
              25869
            ]
          ]
        ]
      },
      "properties": {
        "objectType": "annotation"
      }
    }
  ]
}
```

**Comment by petebankhead** [comment]
Thanks @kaczmarj !

> openslide loader (i think -- i'm not sure how to change this to bioformats)

It's easiest if you create a project - then the docs describe how to specify the library to read the image [here](https://qupath.readthedocs.io/en/0.4/docs/tutorials/projects.html#add-images).

> ```groovy
> // Set number of parallel workers
> qupath.ext.wsinfer.ui.WSInferPrefs.numWorkersProperty().setValue(8);
> ```

Is there any particular reason for choosing 8 - is it based on the number of processors, or testing different values?

**Comment by kaczmarj** [comment]
> Is there any particular reason for choosing 8 - is it based on the number of processors, or testing different values?

the choice of 8 was arbitrary. i have 8 physical cores on my machine, but the main reason is i wanted the value to be set explicitly, because i assume it will have an effect on runtimes.

i get a strange issue when opening the svs file with bioformats. i commented on an existing forum post that describes the same behavior: https://forum.image.sc/t/problem-about-opening-some-svs-slides-in-qupath-v0-3-1-bio-formats-6-8-0/61404/28?u=kaczmarj

**Comment by alanocallaghan** [comment]
> > Is there any particular reason for choosing 8 - is it based on the number of processors, or testing different values?
> 
> the choice of 8 was arbitrary. i have 8 physical cores on my machine, but the main reason is i wanted the value to be set explicitly, because i assume it will have an effect on runtimes.
> 
> i get a strange issue when opening the svs file with bioformats. i commented on an existing forum post that describes the same behavior: https://forum.image.sc/t/problem-about-opening-some-svs-slides-in-qupath-v0-3-1-bio-formats-6-8-0/61404/28?u=kaczmarj

I don't think you should parallelize code by default, if somebody has a 2 core CPU and the scheduler isn't playing nice then that could freeze their system. Or, if you're going to set it by default, I prefer using the number of available cores/threads minus one or two

**Comment by petebankhead** [comment]
> i get a strange issue when opening the svs file with bioformats. i commented on an existing forum post that describes the same behavior: https://forum.image.sc/t/problem-about-opening-some-svs-slides-in-qupath-v0-3-1-bio-formats-6-8-0/61404/28?u=kaczmarj

I saw the thread - can you check https://github.com/qupath/qupath/releases/tag/v0.5.0-rc2 to see if it has already been addressed?

I don't see this issue on my Mac on v0.4.4 or v0.5.0-rc2. The pyramid levels are different from OpenSlide (including 98.8) but the appearance is fine at all levels.

> I don't think you should parallelize code by default, if somebody has a 2 core CPU and the scheduler isn't playing nice then that could freeze their system. Or, if you're going to set it by default, I prefer using the number of available cores/threads minus one or two

I currently have the default set to `min(4, num_processors)` [here](https://github.com/qupath/qupath-extension-wsinfer/blob/main/src/main/java/qupath/ext/wsinfer/ui/WSInferPrefs.java#L42).

Previously, the default was 1 but I found that increasing it would improve performance quite dramatically... up to a limit.

Probably the default should be updated to subtract 1 from the number of processors.

**Comment by MITEL-UNIUD** [comment]
Hi, I did the same test on MacBook Pro 2021 (M1 Pro: 10 CPU cores, 16 GPU cores), 32GB unified memory, MacOS Sonoma 14.5, QuPath 0.5.1, WSInfer 0.30.  I suppose openslide loader since Bio-formats is not yet supporting Apple Silicon.
I also played with batch size by adding this line:
`qupath.ext.wsinfer.ui.WSInferPrefs.batchSizeProperty().setValue(16);`.
Running times:
- **mps** (GPU): 68 seconds (the best was 59, but I was not able to reproduce it). This can be obtained with different workers and batch sizes (up to 24-32), +/- some second.
- **CPU**: it seems to accept much larger batch sizes (tested up to 256). Best time 7'40" (10 workers, 256 batch size). 8 workers and default batch size: 12'07".

**Comment by petebankhead** [comment]
Thanks!

> I suppose openslide loader since Bio-formats is not yet supporting Apple Silicon.

This only affects a couple of file formats (at least .ndpi and .czi). The rest should work on Apple Silicon. With the image open in QuPath, you can check the 'Server type' under the 'Image' tab in QuPath to convert if Bio-Formats or OpenSlide was used.
